### PR TITLE
feat: Enforce `NEXT_PUBLIC_` env key on client with TS

### LINKED ISSUE #1369: feat: Migrate `env.mjs` to `.ts`
### Is your feature request related to a problem? Please describe.

It currently doesn't stop us from defining non-`NEXT_PUBLIC_` key as part of the client env.

### Describe the solution you'd like to see

If we migrate the `.env.mjs` to `.env.ts`, we can utilize the `satisfies` keyword to enforce the ``NEXT_PUBLIC_${string}`` as the client env key.
 
However, I'm actually pretty curious why `.mjs` for `env.mjs`? It would be good if someone can answer this🙏

I personally have the working solution in my personal project, and I am more than happy to help.

### Describe alternate solutions

-

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by chungweileong94**
Closes #1369

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

Enforce `NEXT_PUBLIC_` env key on client, using other non-`NEXT_PUBLIC_` key on client will give you a type error.

---

## Screenshots

image

💯

### CODE REVIEW

File: cli/template/base/src/env.mjs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// @ts-check
```
Comment (juliusmarminge) [inline cli/template/base/src/env.mjs:1]: No need for these as we use `checkJs` in tsconfig

Comment (bai) [inline cli/template/base/src/env.mjs:1]: Do we need ts-check in JS files, given that [we have checkJs set to true in tsconfig](https://github.com/t3-oss/create-t3-app/blob/next/tsconfig.json#L11)?

Comment (chungweileong94) [reply cli/template/base/src/env.mjs:1]: Yeah, I was thinking the same, but it wasn't working on VSCode for some reason. it's working now after restarting the instance.
Thx for pointing that out!

Comment (chungweileong94) [reply cli/template/base/src/env.mjs]: Yeah, you are right. it was some issue with my VSCode earlier, thx for pointing that out!

**Review by juliusmarminge — APPROVED** [review]

**Pull request merged**